Sources/NIOExtras/RequestResponseHandlers+State.swift

```diff
@@ -0,0 +1,250 @@
+    }
+
+    enum DeactiveChannelAction {
+        case fireInactive  // propagate channelInactive to the context
+        // fail the promises and propagate channelInactive to the context
+        case failPromisesAndFireInactive([PromiseBuffer.Promise])
+        case `return`  // do nothing
+    }
+
+    /// Handle `channelInactive`. Drains the buffer if the state machine is still operational.
+    mutating func deactivateChannel() -> DeactiveChannelAction {
+        switch self.state {
+        case .error:
+            // We failed any outstanding promises when we entered the error state and will fail any
+            // new promises in write.
+            assert(self.promiseBuffer.count == 0)
+            return .fireInactive
+        case .inactive:
+            assert(self.promiseBuffer.count == 0)
+            // This is weird, we shouldn't get this more than once but it's not the end of the world either. But in
+            // debug we probably want to crash.
+            assertionFailure("Received channelInactive on an already-inactive handler")
+            return .return
+        case .operational:
+            self.state = .inactive
+            return .failPromisesAndFireInactive(self.promiseBuffer.removeAll())
+        }
+    }
+
+    enum ErrorCaughtAction {
+        // fail the promises and close the context as we received an error
+        case failPromisesAndCloseContext([PromiseBuffer.Promise])
+        case `return`  // do nothing
+    }
+
+    /// Handle `errorCaught`. Transitions to `.error` exacly once as there is no way to make the handler operational again.
+    mutating func errorCaught(error: Error) -> ErrorCaughtAction {
+        guard self.state.isOperational else {
+            assert(self.promiseBuffer.count == 0)
+            return .return
+        }
+        self.state = .error(error)
+        return .failPromisesAndCloseContext(self.promiseBuffer.removeAll())
+    }
+
+    enum ReadPromiseAction {
+        // succeed the returned promise
+        case succeed(PromiseBuffer.Promise)
+        case promiseNotFound  // no matching promise found
+        case bufferEmpty  // buffer is empty
+        case `return`  // ignore (not operational)
+    }
+
+    /// Remove and return the promise for `id` if it is present in the buffer.
+    mutating func readPromise(id: PromiseBuffer.ID) -> ReadPromiseAction {
+        guard self.state.isOperational else {
+            assert(self.promiseBuffer.count == 0)
+            return .return
+        }
+
+        if let promise = self.promiseBuffer.remove(key: id) {
+            return .succeed(promise)
+        } else {
+            if self.promiseBuffer.count == 0 {
+                return .bufferEmpty
+            } else {
+                return .promiseNotFound
+            }
+        }
+    }
+
+    enum WritePromiseAction {
+        case failWith(error: Error)  // do not write, fail promise with error
+        case writeContext  // write the promise to the context
```

Comment:
This is a weird name, we're not writing a context, we're using the context to write a request.